File: dotnet/src/InstrumentComponents/Classes/AsyncPowerMeter.cs
using InstrumentComponents.Dialects;
using InstrumentComponents.Errors;
using InstrumentComponents.Kind;
using InstrumentComponents.Session;
using InstrumentComponents.Scpi;

namespace InstrumentComponents.Classes;

public sealed class AsyncPowerMeter
{
    private readonly AsyncInstrumentSession _session;

    public AsyncPowerMeter(AsyncInstrumentSession session) => _session = session;

    public AsyncInstrumentSession Session => _session;

    private DialectProfile Dialect => _session.DialectFor(InstrumentKind.PowerMeter);

    private string RequireCommand(string key) =>
        Dialect.Command(key) ?? throw new InstrumentUnsupportedException($"power meter dialect missing command '{key}'");

    private string RequireFormatted(string key, params (string Name, string Value)[] vars) =>
        Dialect.FormatCommand(key, vars) ?? throw new InstrumentUnsupportedException($"power meter dialect missing command '{key}'");

    public async Task ConfigureMeasurementAsync(
        PowerUnit unit,
        bool autoRange,
        bool autoAverage,
        double? correctionFreqHz = null,
        double? offsetDb = null,
        CancellationToken cancellationToken = default)
    {
        await _session.Scpi.WriteAsync(RequireFormatted("unit", ("unit", unit.ScpiName())), cancellationToken).ConfigureAwait(false);
        await _session.Scpi.WriteAsync(RequireFormatted("auto_range", ("state", autoRange ? "ON" : "OFF")), cancellationToken).ConfigureAwait(false);
        await _session.Scpi.WriteAsync(RequireFormatted("auto_average", ("state", autoAverage ? "ON" : "OFF")), cancellationToken).ConfigureAwait(false);
        if (correctionFreqHz is { } hz)
            await _session.Scpi.WriteAsync(RequireFormatted("correction_frequency", ("hz", ScpiFormat.Double(hz))), cancellationToken).ConfigureAwait(false);
        if (offsetDb is { } db)
            await _session.Scpi.WriteAsync(RequireFormatted("offset", ("db", ScpiFormat.Double(db))), cancellationToken).ConfigureAwait(false);
    }

    public Task InitiateAsync(CancellationToken cancellationToken = default) =>
        _session.Scpi.WriteAsync(RequireCommand("initiate"), cancellationToken);

    public async Task<double> FetchAsync(CancellationToken cancellationToken = default)
    {
        var resp = await _session.Scpi.QueryAsync(RequireCommand("fetch"), cancellationToken).ConfigureAwait(false);
        return ScpiSession.ParseF64(resp);
    }

    public async Task<double> ReadAsync(CancellationToken cancellationToken = default)
    {
        var resp = await _session.Scpi.QueryAsync(RequireCommand("read"), cancellationToken).ConfigureAwait(false);
        return ScpiSession.ParseF64(resp);
    }
}

File: crates/instrument/src/classes/async_power_meter.rs
use super::dialect_io;
use super::power_meter::PowerUnit;
use instrument_core::error::Result;
use instrument_core::kind::InstrumentKind;
use instrument_core::scpi::AsyncScpiSession;
use instrument_core::AsyncInstrumentSession;

/// Async RF / microwave power meter session view.
pub struct AsyncPowerMeter {
    session: AsyncInstrumentSession,
}

impl AsyncPowerMeter {
    pub fn new(session: AsyncInstrumentSession) -> Self {
        Self { session }
    }

    pub fn session(&self) -> &AsyncInstrumentSession {
        &self.session
    }

    pub fn session_mut(&mut self) -> &mut AsyncInstrumentSession {
        &mut self.session
    }

    fn dialect(&self) -> &'static instrument_core::DialectProfile {
        self.session.dialect_for(InstrumentKind::PowerMeter)
    }

    async fn write_dialect(&mut self, key: &str, vars: &[(&str, String)]) -> Result<()> {
        let cmd = dialect_io::formatted(self.dialect(), key, vars)?;
        self.session.scpi_mut().write(&cmd).await
    }

    async fn query_dialect(&mut self, key: &str) -> Result<String> {
        let cmd = dialect_io::command(self.dialect(), key)?;
        self.session.scpi_mut().query(cmd).await
    }

    /// Configures units, auto-range/average, optional correction frequency and offset.
    pub async fn configure_measurement(
        &mut self,
        unit: PowerUnit,
        auto_range: bool,
        auto_average: bool,
        correction_freq_hz: Option<f64>,
        offset_db: Option<f64>,
    ) -> Result<()> {
        let on_off = |enabled: bool| if enabled { "ON" } else { "OFF" };
        self.write_dialect("unit", &[("unit", unit.scpi_name().into())])
            .await?;
        self.write_dialect("auto_range", &[("state", on_off(auto_range).into())])
            .await?;
        self.write_dialect("auto_average", &[("state", on_off(auto_average).into())])
            .await?;
        if let Some(hz) = correction_freq_hz {
            self.write_dialect("correction_frequency", &[("hz", dialect_io::f64_text(hz))])
                .await?;
        }
        if let Some(db) = offset_db {
            self.write_dialect("offset", &[("db", dialect_io::f64_text(db))])
                .await?;
        }
        Ok(())
    }

    /// Initiates a measurement (INIT).
    pub async fn initiate(&mut self) -> Result<()> {
        let cmd = dialect_io::command(self.dialect(), "initiate")?;
        self.session.scpi_mut().write(cmd).await
    }

    /// Fetches the last initiated measurement (FETC?).
    pub async fn fetch(&mut self) -> Result<f64> {
        AsyncScpiSession::parse_f64(&self.query_dialect("fetch").await?)
    }

    /// Reads a measurement immediately (READ?).
    pub async fn read(&mut self) -> Result<f64> {
        AsyncScpiSession::parse_f64(&self.query_dialect("read").await?)
    }
}
